### ISSUE TITLE
Improved docs for custom fieldtype richtext field using tiptap

### ISSUE THREAD

**Issue by katerlouis** [storyblok/storyblok#887]
**The feature would affect:** (check one with "x")   
- [x] *app.storyblok.com (CMS - Interface)*   
- [ ] *api.storyblok.com (CMS - Content Delivery API)*
- [ ] *mapi.storyblok.com (CMS - Management API)*
- [ ] *a.storyblok.com (CMS - Image Service)*
- [ ] *Other* 

**Is your feature request related to a problem? Please describe.**
On [this page](https://www.storyblok.com/docs/richtext-field) the docs clearly imply that a custom version of a richtext field can be built using tiptap. See screenshot below. But the docs stop here and don't explain how to go from here. I can only assume and tried a lot, without success.
![image](https://user-images.githubusercontent.com/1983382/213397231-ae0ee95b-9d95-4f04-8553-f4630970bdf8.png

**Describe the solution you'd like**
At least one paragraph below the info box on how to get started. A headline, a oneliner paragraph and one code block with a working, minimal field-plugin.

**Comment by SebbeJohansson** [comment]
I want this too! For example sup and sub-script is not available right now.

**Comment by jamesrweb** [comment]
Yeah, as [mentioned](https://github.com/storyblok/storyblok/issues/422#issuecomment-1441703287) in #422, this is quite important to document and get right. Currently there are a lot of open questions that the documentation fails far short of the expected standard of a production tool like Storyblok in this area.

**Comment by SebbeJohansson** [comment]
Extending the tiptap should probebly be a future live stream from storyblok.

**Comment by SebbeJohansson** [comment]
> > Extending the tiptap should probebly be a future live stream from storyblok.
> 
> Great idea, but just to clarify: I'm not talking about extending the existing tiptap field, but creating your own custom one, which the docs state is fairly easy. I got it to work since then with some minor caveats, but the flexibility we get is worth it.
@katerlouis, I would say the documentation is bad for both of those situations. 😁

**Comment by jamesrweb** [comment]
> > Extending the tiptap should probebly be a future live stream from storyblok.
> 
> Great idea, but just to clarify: I'm not talking about extending the existing tiptap field, but creating your own custom one, which the docs state is fairly easy. I got it to work since then with some minor caveats, but the flexibility we get is worth it.

Have you got a working example of how you did this in Storyblok? I found no documentation on this. You mean you replaced the default rich text editor in the dashboard and have full customisation unlocked?

**Comment by jamesrweb** [comment]
> > > Extending the tiptap should probebly be a future live stream from storyblok.
> > 
> > 
> > Great idea, but just to clarify: I'm not talking about extending the existing tiptap field, but creating your own custom one, which the docs state is fairly easy. I got it to work since then with some minor caveats, but the flexibility we get is worth it.
> 
> @katerlouis, I would say the documentation is bad for both of those situations. 😁

Exactly, it is very bad for both cases either way and should be addressed by the storyblok team ASAP.

**Comment by katerlouis** [comment]
> > > Extending the tiptap should probebly be a future live stream from storyblok.
> > 
> > 
> > Great idea, but just to clarify: I'm not talking about extending the existing tiptap field, but creating your own custom one, which the docs state is fairly easy. I got it to work since then with some minor caveats, but the flexibility we get is worth it.
> 
> Have you got a working example of how you did this in Storyblok? I found no documentation on this. You mean you replaced the default rich text editor in the dashboard and have full customisation unlocked?

I did and am currently working on it. It's more work than I expected to be honest. All the flexibility comes at a cost. And a huge disadvantage is not being able to "break out of the iframe" in which the field type plugin lives. Why do I want that? Well.. storybloks Link-Dialog cannot be used compared to how one can use the `sb-asset-selector` – so I have to make my own Link-Dialog and... yeah... 

My work in progress so far, probably won't be of much use without context though
<img width="315" alt="image" src="https://user-images.githubusercontent.com/1983382/221357352-ed6c8863-c611-41d8-8446-335d05a3d9ea.png">

Getting there is actually easier than you think: I created a field plugin with local dev mode as explained [here](https://www.storyblok.com/docs/plugins/enable-local-dev-mode) and installed tiptap like I normally would in any other Vue environment. Yes, this means tiptap will be baked in the exported field plugin, ... but it works :D

**Comment by lisilinhart** [comment]
Hi @katerlouis, thank you for this suggestion. We're planning to make improvements to the richtext field and adding some more extensions like sub and superscript. Can you tell me the reason what you're missing the in the current richtext field, that makes you want to build your own field?

**Comment by katerlouis** [comment]
> Hi @katerlouis, thank you for this suggestion. We're planning to make improvements to the richtext field and adding some more extensions like sub and superscript, which is already in development. Can you tell me the reason what you're missing the in the current richtext field, that makes you want to build your own field?

We needed a lot more from our richtext field, but most of it is arguably highly subjective. What I think everybody would benefit from:

* different kinds of unordered lists (basically classes inside a dropdown for ul)
* different kinds of ordered lists (basically classes inside a dropdown for ol)
* different text sizes 
* coloring / highlighting / marking text sections

I hear you say: "Have you seen 'class options' in the fields settings?" to which I say: yes, I have; but we needed different classes PER NODE; sharing classes between headlines, paragraphs, links and lists is just not realistic and introduces a lot potential for human error; 

That being said: I'm certainly just not the target group for an all-in-one solution richtext field. Thats why I was so happy to read about the ability to just make my own richtext field, which was quickly followed by disappointment due to a lack of documentation on how to get started.

We've solved almost all issues since then; we even built our own custom link-dialog, that unfortunately had to fit inside the tiny space provided by the iframe of the field plugin. 

For those interested what we came up with:
![image](https://user-images.githubusercontent.com/1983382/223405967-a1ca8b8d-4430-4910-b23c-b47e8935b132.png)
![image](https://user-images.githubusercontent.com/1983382/223406002-d028f2f8-8039-4c10-83e2-ae956ef2a457.png)

**Comment by lisilinhart** [comment]
Thanks for the insight @katerlouis. I understand that our richtext solution might not fit everyone, but the goal is to make it as universal usable as possible. You make some good points though about needing a custom solution and I think it makes even more sense, if the content is not within a website, but maybe a mobile framework. That said, building a custom richtext editor is quite complex and I would not advise it to most clients, just because of the sheer amount of implementation work. My hope is that in the future we can build some kind of bridge for our clients to build their own extensions through https://tiptap.dev/guide/custom-extensions and then use that within our richtext editor.

**Comment by SebbeJohansson** [comment]
@lisilinhart Thanks for the update regarding sup and subscript! looking forward to that being in the editor.
Are you looking at adding anything else in the near future? For us the most important thing would be adding color to parts of the text.

**Comment by katerlouis** [comment]
> Thanks for the insight @katerlouis. I understand that our richtext solution might not fit everyone, but the goal is to make it as universal usable as possible. You make some good points though about needing a custom solution and I think it makes even more sense, if the content is not within a website, but maybe a mobile framework. That said, building a custom richtext editor is quite complex and I would not advise it to most clients, just because of the sheer amount of implementation work. My hope is that in the future we can build some kind of bridge for our clients to build their own extensions through https://tiptap.dev/guide/custom-extensions and then use that within our richtext editor.

Thats a fair point; but that doesn't justify the lack of documentation on HOW to get started with a custom tiptap field after mentioning that it's possible. 

So my suggestion (or my request even) would be to either remove this part of the docs, or flesh it out more. 

If you decide to flesh it out: Why not provide a working boiler-plate with the bare minimum of functions, or maybe even the entire functionality of the storyblok tiptap field so that the developer can just alter that in a custom field?

**Comment by SebbeJohansson** [comment]
I had some time left over today, so I spent an hour or so trying to get this to work, but I wasn't successful. 
I used a combination of the following:
https://www.storyblok.com/docs/plugins/introduction#local-development
https://www.storyblok.com/tp/react-field-type-plugin
https://markus.oberlehner.net/blog/building-a-custom-storyblok-field-type-plugin-with-vue/
https://tiptap.dev/installation/vue2

Issues I was facing:
* I couldn't get storyblok to connect to my local environment over HTTPS (resulting in nothing loading).
![image](https://user-images.githubusercontent.com/5083273/223509286-55e70173-5c50-4121-bf57-b9ede4f28f0c.png)
* I tried building locally and uploading the resulting code as described, resulting in an empty plugin without the editor.
![image](https://user-images.githubusercontent.com/5083273/223509343-797d4ec0-710c-419d-8b66-d9a6714cb66c.png)

My attempt can be found here:
https://github.com/SebbeJohansson/storyblok-custom-tiptap-plugin/blob/main/src/Plugin.vue

@lisilinhart Maybe I could get some help from the extension team to look at this?

**Comment by katerlouis** [comment]
I recommend holding the editor instance inside `data()` instead of storyblok plugins `model`. The editor itself does not need to be stored. It's unlikely to be the reason for your issue, but just to be sure, I'd put it in `data()` – you want to save the output of tiptap in `model.value`. 

To do that, I created a `TipTap.vue` file in order to use custom v-model.

Plugin.vue
```vue
<template>
  <div class="dumbo-richtext">
    <TipTap v-model="model.content" />
  </div>
</template>

<script>
import TipTap from './TipTap.vue';

export default {

  components: {
    TipTap,
  },

  mixins: [window.Storyblok.plugin],

  watch: {
    'model': {
      handler: function(value) {
        this.$emit('changed-model', value);
      },
      deep: true,
    },
  },

  methods: {
    initWith() {
      return {
        plugin: 'dumbo-tiptap',
        content: '',
      };
    },

    pluginCreated() {
      console.log('plugin:created dumbos tip tap');
    },
  },
};
</script>
```

TipTap.vue
```vue
<template>
  <div class="tip-tap">
    <TipTapMenu
      v-if="editor"
      :editor="editor"
    />
    <TipTapBubble
      v-if="editor"
      :editor="editor"
    />
    <EditorContent :editor="editor" class="content" />
  </div>  
</template>

<script>
import TipTapBubble from './TipTapBubble.vue'; // my custom bubble menu
import TipTapMenu from './TipTapMenu.vue'; // my custom menu

import { Editor, EditorContent } from '@tiptap/vue-2';
import StarterKit from '@tiptap/starter-kit';

import debounce from 'lodash.debounce';

export default {
  name: 'TipTap',

  components: {
    EditorContent,
    TipTapMenu,
    TipTapBubble,
  },

  props: {
    // for custom v-model
    value: {
      type: undefined,
      required: true,
    },
  },

  data() {
    return {
      editor: null,
    };
  },

  mounted() {
    this.editor = new Editor({
      extensions: [
        StarterKit.configure({
          // see tiptap docs
        }),
      ],

      content: this.value, 
    });

    // (!)
    this.editor.on('update', debounce(({ editor }) => {
      this.$emit('input', editor.getJSON());
    }, 200));
  },
};
</script>
```

**Comment by lisilinhart** [comment]
> > Thanks for the insight @katerlouis. I understand that our richtext solution might not fit everyone, but the goal is to make it as universal usable as possible. You make some good points though about needing a custom solution and I think it makes even more sense, if the content is not within a website, but maybe a mobile framework. That said, building a custom richtext editor is quite complex and I would not advise it to most clients, just because of the sheer amount of implementation work. My hope is that in the future we can build some kind of bridge for our clients to build their own extensions through https://tiptap.dev/guide/custom-extensions and then use that within our richtext editor.
> 
> Thats a fair point; but that doesn't justify the lack of documentation on HOW to get started with a custom tiptap field after mentioning that it's possible.
> 
> So my suggestion (or my request even) would be to either remove this part of the docs, or flesh it out more. To me, and I know thats subjective, theres not much more frustrating in the dev world than ambiguous misleading, outdated or just false docs. We all work on multiple projects, with varying tech stacks. The web is driven by third party tools, using other third party tools etc.– Good documentation is vital these days.
> 
> If you decide to flesh it out: Why not provide a working boiler-plate with the bare minimum of functions, or maybe even the entire functionality of the storyblok tiptap field so that the developer can just alter that in a custom field?

@katerlouis  I agree on the documentation point. I scheduled a meeting with our documentation team to get their point of view on this topic and to align all the different people involved in the richtext development. I generally think the documentation on richtext needs some doover from our side. We also have a few of the suggestions that are missing in richtext already in the development phase and coming soon, including:
- sub & sup
- text color and text highlight
- undo / redo
- emoji

**Comment by SebbeJohansson** [comment]
@lisilinhart Amazing. That sounds great 🥳

**Comment by alvarosabu** [comment]
Hi @katerlouis, we are closing this issue because this repository was used in the early days for support-related questions, but we've since transitioned to more effective channels for helping our users.

Please, refer to the following resources:

Storyblok Github Org: https://github.com/storyblok
Storyblok Docs: https://www.storyblok.com/docs
Official Support: https://support.storyblok.com
Join our Discord: https://www.storyblok.com/join-discord

Thanks for the understanding 🙏🏻
